### PR TITLE
Add optional wrappers to AOT headers to allow const input buffers (Issue #3077)

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

**Comment by steven-johnson** [comment]
clean, PTAL

**Comment by abadams** [comment]
Currently exploring the alternative of just forking out the bounds query entrypoint. If that works we can just mark the main entrypoint as taking const halide_buffer_t args.

We could even add the const early, as long as we believe the separate entrypoint is going to work out.

**Comment by steven-johnson** [comment]
fair enough.

**Comment by steven-johnson** [comment]
TODO: (from #3191): "We could go back to the const wrapper approach, assert that the buffers is not in bounds query mode before the call, and additionally assert that the device and flags fields didn't change after the call to the pipeline."

**Comment by steven-johnson** [comment]
There's still a problem with this approach: we'd like to be able to seamlessly pass a `const Buffer<T>` or a `Buffer<const T>` as inputs (which this accomplishes), but we'd also like to **forbid** passing either as output (which this does not). To accomplish that we'd need to change the definition of `operator halide_buffer_t*` in HalideBuffer.h to fail to compile when `T` is a const type.

**Comment by abadams** [comment]
a `const Buffer<T> &` or `const halide_buffer_t *` is perfectly fine as an output. Especially the latter - it's a vanilla C struct so vanilla C const rules apply. The former obeys our non-STLy Buffer<T> constness model.

Regarding a `Buffer<const T>` - at one point we considered have a separate halide_buffer_t variant with a const qualifier on the host pointer, but it was more complicated than it's worth. Hrm.

We could also make this wrapper (or yet another wrapper) take and return `Buffer<T>s`?

**Comment by steven-johnson** [comment]
> a const Buffer<T> & or const halide_buffer_t * is perfectly fine as an output

(1) Maybe it's legal, but it's confusing as hell; `const Foo&` as an output is going to look and feel wrong to ~all C++ coders, even if you explain out the wazoo that it's equivalent to `const void**`.
(2) Is that even true? Can't (say) a GPU-compiled function set the device and/or dirty flags on the output?

**Comment by steven-johnson** [comment]
> We could also make this wrapper (or yet another wrapper) take and return Buffer<T>s?

I've pondered that. It might be interesting. Inside Google have quite a mixture of code using both halide_buffer_t and Halide::Runtime::Buffer so it might be challenging. At the moment I'm resisting it, hoping I can keep the abstraction simpler.

**Comment by steven-johnson** [comment]
Experimenting with the device-and-flags-assert code inside google, there are more fun corner cases: e.g., the case where an input buffer is passed as `const Buffer&` all the way down the call stack... only to be passed as input to an HVX-compiled pipeline, which internally calls `halide_copy_to_device`, which sets the `device` field, which makes the newly-inserted assert fire.

You can (and probably will) argue that this code was always wrong in the first place for doing this, but consider: what if the code chain originated with only cpu-only pipelines (and thus const was right and good), then had HVX added later... is it better to require the code to propagate the nonconstness up the call chain, or to just treat device as (effectively) a `mutable` field? Neither sounds appealing. (And the fact that the assertion-failure will only fire in non-optimized-builds means that we are certainly going to miss some of these cases on our first go-round, as many internal tests run in optimized builds only.)

**Comment by steven-johnson** [comment]
Is it even remotely feasible to try to determine (and enforce) buffer mutability at compiletime, rather than via runtime assertions? (ie: can the Halide compiler determine that the pipeline definitely will/will-not mutate the halide_buffer_t itself, and from that, emit a const-or-not declaration?)

I suspect the answer is 'no', if for no other reason than `define_extern` can do just about anything. (I guess you could be conservative and say that the presence of one of those means it could be mutated, but that's not very satisfying.)

**Comment by abadams** [comment]
I mean, it always *might* mutate every buffer, given bounds queries. We could deduce the condition under which it's mutable, but it would be a runtime Expr.

**Comment by abadams** [comment]
Regarding the `const void **` thing, this is in a machine-generated header, and the option where we include a const generates strictly fewer compiler errors than the option without it. The kinds of people who go digging just for curiosity can have the meaning of `const void **` explained to them.

**Comment by steven-johnson** [comment]
So the addition of the `assert` checks on `device` and `flags` makes this a really bad fit for my particular use case:

- as mentioned above, de facto situation inside google is that the vast majority of 'input' buffers tend to be passed around by const reference; at the point of calling into halide, this is usually `const_cast` into a halide_buffer_t*. If the call happens to mutate flags/device, this is obscured by the casting and the callers essentially end up treating these fields as C++ `mutable`.
- a reasonable C++ coder would point out that buffers passed into such code paths should really be passed by nonconst reference, and they'd be right; unfortunately, C++ at google is not actually reasonable, as nonconst references are essentially forbidden as function arguments (the theory being that "inputs should be passed by constref, outputs should be passed by pointer"). Yes, this is unfortunate, but it's the land we live in here at El Goog.
- so.... passing input-buffers-that-might-have-device-or-flags-mutated by pointer is _technically_ doable, but a major pain and dislocation, if for no other reason that it means explicit temporaries become necessary where they wouldn't be otherwise (e.g. if you have a Buffer<T> that you want to pass to something that takes a Buffer<const T>). 
- alternatively, of course, I could just make a local copy of a const-Buffer when necessary. But it's extra overghead. It's not a *huge* amount of overhead, but it is nonzero. 

So basically I guess I'm proposing that these wrappers treat device/flags as C++ `mutable` -- or maybe add yet another (yuck) preprocessor option to actually enable these assertions or not.

**Comment by abadams** [comment]
When using the raw halide_buffer_t * interface, I don't think being gifted a device allocation that you're responsible for freeing really should fall under the definition of mutable.

When using `Buffer<T>` maybe it's not so bad, because the device allocation gets reference counted.

Therefore how about a const-qualified ::mutable_raw_buffer() method on Buffer<T> that returns a non-const halide_buffer_t, or (this is grosser), have const `Buffer<T>` implicitly cast to non-const halide_buffer_t as a local google modification designed to work around const-incorrect legacy code.

**Comment by steven-johnson** [comment]
> When using the raw halide_buffer_t * interface, I don't think being gifted a device allocation that you're responsible for freeing really should fall under the definition of mutable.

This is a good point -- my use case for this really only involves the `Buffer` case so I wasn't appropriately considering that. Bad engineer, no cookie. (I'm wonder if it would make more sense to just replace this PR entirely with one that emits wrappers that deal with `Buffer` directly; declaring `device` mutable for our purposes is a lot more palatable if you know that the underlying halide_buffer_t has proper lifetime management wrapped around it.)

> Therefore how about a const-qualified ::mutable_raw_buffer() method on Buffer that returns a non-const halide_buffer_t

Hmm... that would probably work, but at the expense of ugly call sites. (Arguably less ugly than `const_cast`, of course.) Let me do some experimenting with it and see how it feels. 

> or (this is grosser), have const Buffer<T> implicitly cast to non-const halide_buffer_t as a local google modification designed to work around const-incorrect legacy code.

That is indeed grosser :-) implicit conversions are ugly enough as-is; the implicit `operator halide_buffer_t*` conversion we have now is handy but still a little scary for my taste.

**Comment by abadams** [comment]
Auto-generating a wrapper that takes mutable refs/pointers to `Buffer<const T>`s for inputs and a mutable ref/pointer to a `Buffer<T>` for the output would be const-correct. A Runtime::Buffer<T> can silently convert to a Runtime::Buffer<const T>, I think, so the `const T` on the inputs is strictly helpful.

Making it additionally permit `const Buffer<const T> &` on the inputs with an assertion that they're not in bounds query mode and a comment that device and flags are considered mutable seems a little dodgy but would solve your problem and be OK with me.

**Comment by steven-johnson** [comment]
> A `Runtime::Buffer<T>` can silently convert to a `Runtime::Buffer<const T>`

Yes, but a `Runtime::Buffer<T> &` does not autoconvert to a `Runtime::Buffer<const T> &`. (No matter, can fix with overloads.)

**Comment by steven-johnson** [comment]
This is likely to be dropped in favor of https://github.com/halide/Halide/pull/3206

**Comment by steven-johnson** [comment]
This is definitely not going to land in its current form; closing
